### DISCUSSION TITLE
detectionMode=true does not warn

### DISCUSSION THREAD

**Discussion by apopaa** [sse-secure-systems/connaisseur#222] [Q&A] [answered]
I have turned on 'detectionMode=true' and run your unsigned testimage example, and I did not get the warning.
So I have 2 questions about detectionMode:
1. Where does Connaisseur log an error message for images that do not comply with the policy?
2. How is flipping the detectionMode flag supposed to work? Do I update its value in helm/values.yaml and ... do a "helm upgrade"? Last time I did that, in 1.5.0, I broke my K8s cluster and you told me to use "helm uninstall"+"helm install again". Am I still to do that now, in 2.1.0? Kind of tedious to always have to uninstall and reinstall, especially for small changes like a flipped flag.

If I turn off "detectionMode=false", the unsigned images do get rejected. So the controller is doing its job, it's just not warning, which makes it risky to deploy this in an existing cluster.

I am using a "kind" cluster and ACR as the registry.

**Comment by xopham** [comment, marked as answer]
Hi @apopaa , to your questions:
1. detectionMode warnings and errors: As mentioned in the [docs here](https://sse-secure-systems.github.io/connaisseur/v2.1.0/features/detection_mode/), there is a direct warning from k8s: 
![image](https://user-images.githubusercontent.com/37963496/127997225-d74f97f5-6649-4e0d-a20f-bf00940258ca.png)
These warnings were introduced as a [feature in Kubernetes v1.19](https://kubernetes.io/blog/2020/09/03/warnings/). Haven't had the time to check whether the warning is issued in Kind as well and which k8s version kind is using.
The logged errors can be found in the pod logs:
```bash
$ kubectl logs -n connaisseur <pod>
[2021-08-03 10:00:01,624] ERROR: {'message': 'Unable to find signed digest for image docker.io/securesystemsengineering/testimage:unsigned.', 'context': {'image_name': 'docker.io/securesystemsengineering/testimage:unsigned', 'detection_mode': True, 'user': 'admin', 'operation': 'CREATE', 'kind': 'Pod', 'name': 'unsigned', 'namespace': 'default', 'image': 'docker.io/securesystemsengineering/testimage:unsigned'}}
[2021-08-03 10:00:01,625] INFO: 192.168.188.96 - - [03/Aug/2021 10:00:01] "POST /mutate?timeout=30s HTTP/1.1" 200 -
```
However, in most cases the warnings are nice when playing around using `kubectl`, but in a real environment, I would recommend to either use error logs. In a typical cluster there should be some sort of logging service to export those logs. Otherwise, one can also use the [alerting feature](https://sse-secure-systems.github.io/connaisseur/v2.1.0/features/alerting/).

2. Updating configuration: Currently, the only way to do this is via your mentioned `helm uninstall` + `helm install` workflow. We are currently investigating how to improve the update workflow, as documented in ticket https://github.com/sse-secure-systems/connaisseur/issues/181

**Reply by apopaa** [reply]
That's the documentation I was reading about detectionMode, but it doesn't say that it is implemented using K8s warnings introduced in v1.19, or else I wouldn't have expected to get them seeing how I was using "kind" version 1.18.6 (to match what I have in production).
After upgrading "kind" to [v1.19.11](https://hub.docker.com/layers/kindest/node/v1.19.11/images/sha256-558780c13d9736d9ba55fa695a5bb8f403a53e42b95bea54fbdc4712181f5e3f?context=explore) I am now getting the warnings.

But more importantly, I found the error entries in the logs of the Connaisseur pods, both in 1.18 and in 1.19, and since that's a better way to check signing I will use that. Thanks for the help.

**Reply by xopham** [reply]
thanks for the feedback :pray:  We'll extend the docs accordingly: #223
